django/db/models/fields/files.py

```diff
@@ -91,7 +91,14 @@ def open(self, mode="rb"):
     def save(self, name, content, save=True):
         name = self.field.generate_filename(self.instance, name)
         self.name = self.storage.save(name, content, max_length=self.field.max_length)
-        setattr(self.instance, self.field.attname, self.name)
+
+        # Set the file-like content using the descriptors __set__ method.
+        setattr(self.instance, self.field.attname, content)
+
+        # Update the name in case generate_filename or the storage backend
+        # changed the name. Avoids re-running the descriptor logic.
+        self.instance.__dict__[self.field.attname] = self.name
+
```

Comment:
It’s great to comment this, but I think the current content slightly overexplains. I propose this:

```suggestion
        setattr(self.instance, self.field.attname, content)
        # Update the name in case generate_filename() or storage.save() changed
        # it, but bypass the descriptor to avoid re-reading the file.
        self.instance.__dict__[self.field.attname] = self.name
```